Cython/Compiler/Nodes.py

```diff
@@ -2072,26 +2072,33 @@ def generate_function_definitions(self, env, code):
             if not entry.used or entry.in_closure:
                 continue
 
-            if entry.type.is_memoryviewslice:
-                code.put_xdecref_memoryviewslice(entry.cname, have_gil=not lenv.nogil)
-            elif entry.type.is_pyobject:
-                if not entry.is_arg or len(entry.cf_assignments) > 1:
-                    if entry.xdecref_cleanup:
-                        code.put_var_xdecref(entry)
-                    else:
-                        code.put_var_decref(entry)
+            if entry.type.is_pyobject:
+                #if not (not entry.is_arg or len(entry.cf_assignments) > 1):
+                if entry.is_arg and len(entry.cf_assignments) <= 1:
+                    continue
+            code.put_var_xdecref(entry,
+                                 do_for_memoryviewslice=True,
+                                 have_gil=not lenv.nogil)
+                        # TODO ideally should pick based on
+                        # entry.xdecref_cleanup but this doesn't seem to be set reliably
+                        # for regular variables
```

Comment:
https://github.com/cython/cython/blob/0d272a51338467a5918911bed068d96bd28ad425/Cython/Compiler/Code.py#L2176-L2179

The current ``put_var_decref` - which has been there for 9 years - has generated `xdecref` whether you called `put_var_decref` or `put_var_xdecref`. So this change doesn't actually degrade the generated code. If definitely crashes sometimes if you try to use `xdecref_cleanup`.

I'll can try to probe a bit deeper and see if there's something obvious missed that'll get the desired behaviour though. But if I succeed it'll be an improvement, rather than restoring the status quo.